src/Utilities/Compiler/CodeMetrics/MetricsHelper.cs

```diff
@@ -280,6 +280,11 @@ private static void AddCoupledNamedTypesCore(ImmutableHashSet<INamedTypeSymbol>.
             // Compat
             static bool isIgnoreableType(INamedTypeSymbol namedType)
```

Comment:
I think that the compiler generated types shall also be excluded:
`namedType.GetAttributes().Any(a => a.AttributeClass.Equals(compilerGeneratedAttributeType) || a.AttributeClass.Equals(generatedCodeAttributeType))`

Reply:
Sounds reasonable to me.

Reply:
@Evangelink I'm working on some unit tests and fixes for what you're describing here. My plan for generics is to include the Unbound generic type (`IEnumerable<>`) in the list of coupled types instead of the actual generic type (`IEnumerable<int>`). Does that sound good?

There's currently an optimization where we avoid recursively checking the generic sub-types if we've already seen the exact generic type we're looking at. Since I won't be adding the actual generic type to the builder anymore, the simplest approach would be to drop that optimization. In theory this could yield degraded performance in code that has a lot of references to nested generic types (`G<G<G<G<a>>>>`), but I think that will be rare. Does that sound all right to you?

Reply:
All right, I've fixed the code in the way I describe above, and added a couple of tests, including one for the specific code I used in #2133 .

The `ConditionalLogic_CoalesceAndConditionalAccess` test previously expected that the generic nullable type (e.g. `bool?`) would be referenced. In keeping with the approach I described, I'm changing that to expect the Unbound generic type instead (e.g. `Nullable<>`, represented as `T?` in the output). But I wanted to check that we want `Nullable<>` to count at all, and not to be treated as an ignored type, the way we ignore `int`, `bool`, etc.? Same goes for `Action` and `Func<>`: we are no longer generating separate coupled types for each generic type of Action and Func, so I consider my original issue fixed regardless, but an unbound generic type is being counted if the code has a lambda or LINQ expression.

In terms of compiler-generated code, it appears that there's nothing showing for closures, so I'm not including the attribute check you mentioned for this pull request. I am curious to see what happens if we use `yield return` or `async`/`await`, but I don't think that should necessarily be something I address in this pull request.